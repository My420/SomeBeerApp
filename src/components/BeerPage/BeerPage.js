import React from 'react';
import PropTypes from 'prop-types';
import { Record } from 'immutable';
import { connect } from 'react-redux';
import BeerItem from '../BeerItem/BeerItem';
import IconLoadingSvg from '../IconLoadingSvg/IconLoadingSvg';
import ErrorMsg from '../ErrorMsg/ErrorMsg';
import styles from './BeerPage.module.scss';
import getBeerById from '../../ActionCreator/getBeerById';
import LoadStatusSwitcher from '../LoadStatusSwitcher/LoadStatusSwitcher';

export class BeerPage extends React.Component {
  componentDidMount() {
    const { getBeerData, id } = this.props;
    getBeerData(id);
  }

  render() {
    const {
      history,
      isLoaded,
      isLoading,
      isError,
      errorMessage,
      data
    } = this.props;

    const LoadingComponent = () => {
      return (
        <div className={styles.loader}>
          <IconLoadingSvg color="#e69e63" />
        </div>
      );
    };

    const ErrorComponent = () => {
      return (
        <div className={styles.error}>
          <ErrorMsg errorMsg={errorMessage} />
        </div>
      );
    };

    const DataComponent = () => {
      return <BeerItem history={history} data={data} />;
    };

    return (
      <React.Fragment>
        <LoadStatusSwitcher
          isLoading={isLoading}
          isError={isError}
          isLoaded={isLoaded}
          LoadingComponent={LoadingComponent}
          ErrorComponent={ErrorComponent}
          DataComponent={DataComponent}
        />
      </React.Fragment>
    );
  }
}

const mapStateToProps = store => {
  return {
    isLoaded: store.beerItem.isLoaded,
    isLoading: store.beerItem.isLoading,
    isError: store.beerItem.isError,
    errorMessage: store.beerItem.errorMessage,
    data: store.beerItem.data
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getBeerData: id => {
      dispatch(getBeerById(id));
    }
  };
};

BeerPage.propTypes = {
  id: PropTypes.string.isRequired,
  isLoaded: PropTypes.bool.isRequired, // form connect
  isLoading: PropTypes.bool.isRequired, // form connect
  isError: PropTypes.bool.isRequired, // form connect
  errorMessage: PropTypes.string.isRequired, // form connect
  data: PropTypes.instanceOf(Record).isRequired, // form connect
  getBeerData: PropTypes.func.isRequired // form connect
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BeerPage);
