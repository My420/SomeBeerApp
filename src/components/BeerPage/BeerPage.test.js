import React from 'react';
import { shallow } from 'enzyme';
import { BeerPage } from './BeerPage';

describe('test <BeerPage /> component', () => {
  const mockFunc = jest.fn();
  const props = {
    id: 10,
    history: 'history',
    isLoaded: true,
    isLoading: false,
    isError: false,
    errorMessage: '',
    data: 'test data',
    getBeerData: mockFunc
  };

  const wrapper = shallow(<BeerPage {...props} />);
  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('getBeerData prop func should be called in compDidMount with correct arg', () => {
    expect(mockFunc.mock.calls.length).toBe(1);
    expect(mockFunc.mock.calls[0][0]).toBe(props.id);
  });

  test('should render LoadStatusSwitcher component with correct props', () => {
    const { isLoaded, isLoading, isError } = props;
    expect(wrapper.find('LoadStatusSwitcher')).toHaveLength(1);
    expect(wrapper.find('LoadStatusSwitcher').prop('isLoading')).toBe(
      isLoading
    );
    expect(wrapper.find('LoadStatusSwitcher').prop('isLoaded')).toBe(isLoaded);
    expect(wrapper.find('LoadStatusSwitcher').prop('isError')).toBe(isError);
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('LoadingComponent')
    ).toBe('function');
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('ErrorComponent')
    ).toBe('function');
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('DataComponent')
    ).toBe('function');
  });
});
