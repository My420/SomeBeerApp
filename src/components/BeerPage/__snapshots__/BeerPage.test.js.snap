// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <BeerPage /> component should match snapshot 1`] = `
<Fragment>
  <LoadStatusSwitcher
    DataComponent={[Function]}
    ErrorComponent={[Function]}
    LoadingComponent={[Function]}
    isError={false}
    isLoaded={true}
    isLoading={false}
  />
</Fragment>
`;
